impl/src/main/java/org/ehcache/internal/store/tiering/CacheStore.java

```diff
@@ -0,0 +1,350 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.internal.store.tiering;
+
+import org.ehcache.Cache;
+import org.ehcache.events.StoreEventListener;
+import org.ehcache.exceptions.CacheAccessException;
+import org.ehcache.function.BiFunction;
+import org.ehcache.function.Function;
+import org.ehcache.function.NullaryFunction;
+import org.ehcache.internal.store.OnHeapStore;
+import org.ehcache.internal.store.disk.DiskStore;
+import org.ehcache.spi.ServiceProvider;
+import org.ehcache.spi.cache.Store;
+import org.ehcache.spi.cache.tiering.AuthoritativeTier;
+import org.ehcache.spi.cache.tiering.CachingTier;
+import org.ehcache.spi.service.ServiceConfiguration;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author Ludovic Orban
+ */
+public class CacheStore<K, V> implements Store<K, V> {
+
+  private final CachingTier<K, V> cachingTier;
+  private final AuthoritativeTier<K, V> authoritativeTier;
+
+  public CacheStore(CachingTier<K, V> cachingTier, AuthoritativeTier<K, V> authoritativeTier) {
+    this.cachingTier = cachingTier;
+    this.authoritativeTier = authoritativeTier;
+
+    this.cachingTier.addInvalidationListener(new CachingTier.InvalidationListener<K, V>() {
```

Comment:
I think this code path is racy. But for it being called from the constructor. 
So I'd either make it thread safe, or support a single listener to ever be registered (and can't be changed). But right now this `add` method (and the remove too, though it isn't used) scare me, given their current implementation.